### PR TITLE
Review for DM-981

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h [hunk 1]
```diff
@@ -1,46 +1,154 @@
...
+#include "lsst/meas/base/CentroidUtilities.h"
+#include "lsst/meas/base/FlagHandler.h"
+#include "lsst/meas/base/InputUtilities.h"
 
 namespace lsst { namespace meas { namespace extensions { namespace shapeHSM {
 
-class HsmSourceMomentsControl : public algorithms::ShapeControl {
+// This is here just as a parent class for the following
+class HsmMomentsControl {
+};
```
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h:38]: What purpose does this serve as a base class if it's empty?

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h [hunk 2]
```diff
@@ -1,46 +1,154 @@
...
 public:
-    HsmPsfMomentsControl() : ShapeControl("shape.hsm.psfMoments") {}
-    HsmPsfMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {}
-    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) : algorithms::ShapeControl(other) {}
+    /// A typedef to the Control object for this algorithm, defined above.
+    ///  The control object contains the configuration parameters for this algorithm.
+    typedef HsmMomentsControl Control;
+
+    /// @brief Initialize with standard field names and customized documentation.
+    HsmMomentsAlgorithm(Control const & ctrl, std::string const & name, afw::table::Schema & schema, char const* doc) :
```
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h]: Given that the constructor doesn't actually use the control object, and this class is only used as a base class, I think you can remove both the control object argument and the typedef to it in this class.

Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h]: Also, this line is too long (but I think removing the control argument will fix that).

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h [hunk 3]
```diff
@@ -1,46 +1,154 @@
...
+    template<typename PixelT>
+    void calculate(
+        afw::table::SourceRecord& source, // Source for recording moments
+        PTR(afw::image::Image<PixelT>) const& afwImage, // Image on which to measure moments
+        PTR(afw::image::Mask<afw::image::MaskPixel>) const& afwMask, // Mask for image
+        afw::geom::Box2I const& bbox,     // Bounding box
+        afw::geom::Point2D const& center, // Starting center for measuring moments
+        afw::image::MaskPixel const badPixelMask, // Bitmask for bad pixels
+        float const width            // PSF width estimate, for starting moments
+        ) const;
```
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h]: Now that this is a public class, these parameter comments should be upgraded to full Doxygen comments (either use "//<" or move them to "@param" lines in the function Doxygen.

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h [hunk 4]
```diff
@@ -1,46 +1,154 @@
...
 private:
-    virtual PTR(algorithms::AlgorithmControl) _clone() const;
-    virtual PTR(algorithms::Algorithm) _makeAlgorithm(
-        afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
+    Control _ctrl;
+    // These methods are defined private to prevent shadowing of the parent classes
+    void measure(
+        afw::table::SourceRecord & measRecord,
+        afw::image::Exposure<float> const & exposure
+    ) const;
```
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h:114]: Just a heads-up: @jdswinbank's DM-2131 (also in review) changing how we work around the problem with shadowed `measure` methods - after that's merged, these will no longer have to be private in derived classes.  So you may want to keep an eye on that, and if it gets merged first, move this to public.  (If yours gets merged first, it'll be his job to update shapeHSM, I think).

File: include/lsst/meas/extensions/shapeHSM/HsmShapeControl.h [hunk 5]
```diff
@@ -11,62 +38,147 @@ enum MeasType {
...
 
 class HsmShapeBjControl : public HsmShapeControl {
 public:
-    HsmShapeBjControl() : HsmShapeControl("shape.hsm.bj") {}
-private:
-    virtual PTR(algorithms::AlgorithmControl) _clone() const;
-    virtual PTR(algorithms::Algorithm) _makeAlgorithm(
-        afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
-    ) const;
+    HsmShapeBjControl() : HsmShapeControl("extensions_shapeHSM_HsmShapeBj") {}
```
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmShapeControl.h]: As I recommended with Kron, let's shorted "extensions" to just "ext" in these names.

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h (hunk 3 above)
Comment (TallJimbo) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h]: And you should also align the close-paren with the first line of the function.

File: python/lsst/meas/extensions/shapeHSM/__init__.py [hunk 6]
```diff
@@ -23,11 +23,17 @@
...
+lsst.meas.base.wrapSimpleAlgorithm(HsmShapeLinearAlgorithm, name="ext_shapeHSM_HsmShapeLinear" 
+    ,Control=HsmShapeLinearControl, executionOrder=3.0)
+lsst.meas.base.wrapSimpleAlgorithm(HsmShapeKsbAlgorithm, name="ext_shapeHSM_HsmShapeKsb" 
+    ,Control=HsmShapeKsbControl, executionOrder=3.0)
+lsst.meas.base.wrapSimpleAlgorithm(HsmShapeRegaussAlgorithm, name="ext_shapeHSM_HsmShapeRegauss" 
+    ,Control=HsmShapeRegaussControl, executionOrder=3.0)
+lsst.meas.base.wrapSimpleAlgorithm(HsmSourceMomentsAlgorithm, name="ext_shapeHSM_HsmMoments" 
+    ,Control=HsmSourceMomentsControl, executionOrder=3.0)
+lsst.meas.base.wrapSimpleAlgorithm(HsmMomentsAlgorithm, name="ext_shapeHSM_HsmPsfMoments" 
+    ,Control=HsmPsfMomentsControl, executionOrder=3.0)
```
Comment (TallJimbo) [inline python/lsst/meas/extensions/shapeHSM/__init__.py]: All these lines that start with a comma should have the comma on the previous line.

File: src/HsmMoments.cc [hunk 7]
```diff
@@ -172,18 +97,20 @@ void HsmSourceMoments::_apply(
 
     double const psfSigma = exposure.getPsf()->computeShape(center).getTraceRadius();
 
-    HsmMoments::calculate(source, exposure.getMaskedImage().getImage(), exposure.getMaskedImage().getMask(),
+    HsmMomentsAlgorithm::calculate(source, exposure.getMaskedImage().getImage(), exposure.getMaskedImage().getMask(),
```
Comment (TallJimbo) [inline src/HsmMoments.cc]: Line is too long.

File: src/HsmMoments.cc [hunk 8]
```diff
@@ -172,18 +97,20 @@ void HsmSourceMoments::_apply(
...
                           bbox, center, badPixelMask, 2.5*psfSigma);
 }
 
+void HsmSourceMomentsAlgorithm::fail(afw::table::SourceRecord & measRecord, base::MeasurementError * error) const {
```
Comment (TallJimbo) [inline src/HsmMoments.cc]: Also too long.

File: src/HsmMoments.cc [hunk 9]
```diff
@@ -194,33 +121,11 @@ void HsmPsfMoments::_apply(
     mask->setXY0(image->getXY0());
 
     double const psfSigma = exposure.getPsf()->computeShape(center).getTraceRadius();
-    HsmMoments::calculate(source, image, mask, image->getBBox(afw::image::PARENT), afw::geom::Point2D(0, 0),
+    HsmMomentsAlgorithm::calculate(source, image, mask, image->getBBox(afw::image::PARENT), afw::geom::Point2D(0, 0),
                           0, psfSigma);
```
Comment (TallJimbo) [inline src/HsmMoments.cc]: Too long.  Should also fix indentation of second line.

File: src/HsmMoments.cc [hunk 10]
```diff
@@ -194,33 +121,11 @@ void HsmPsfMoments::_apply(
...
-
-PTR(algorithms::AlgorithmControl) HsmSourceMomentsControl::_clone() const {
-    return boost::make_shared<HsmSourceMomentsControl>(*this);
-}
-
-PTR(algorithms::Algorithm) HsmSourceMomentsControl::_makeAlgorithm(
-    afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
-) const {
-    return boost::make_shared<HsmSourceMoments>(*this, boost::ref(schema), "source adaptive moments from HSM");
+void HsmPsfMomentsAlgorithm::fail(afw::table::SourceRecord & measRecord, base::MeasurementError * error) const {
```
Comment (TallJimbo) [inline src/HsmMoments.cc]: Too long.

File: src/HsmShape.cc [hunk 11]
```diff
@@ -197,10 +173,9 @@ void HsmShape::_apply(
     *dummyMask = 1;
     ImageConverter<int> const psfMask(dummyMask);
 
-    // Calculate the sky variance
     afw::math::StatisticsControl sctrl;
     sctrl.setAndMask(badPixelMask);
-    typename ExposureT::MaskedImageT::Variance const variance(*exposure.getMaskedImage().getVariance(), bbox,
+    afw::image::Exposure<float>::MaskedImageT::Variance const variance(*exposure.getMaskedImage().getVariance(), bbox,
```
Comment (TallJimbo) [inline src/HsmShape.cc]: Too long.

File: tests/testHsm.py [hunk 12]
```diff
@@ -174,7 +192,6 @@ def runMeasurement(self, algorithmName, imageid, x, y):
                                          afwGeom.Point2D(1.0,1.0),
```
Comment (TallJimbo) [inline tests/testHsm.py:192]: I know this one isn't your fault, but could you fix this too-long line too?

File: tests/testHsm.py [hunk 13]
```diff
@@ -288,32 +306,30 @@ def testHsmPsfMoments(self):
...
-
-            moments = source.get("shape.hsm.psfMoments")
-            centroid = source.get("shape.hsm.psfMoments.centroid")
-
-            self.assertAlmostEqual(centroid[0], 0.0, 3)
-            self.assertAlmostEqual(centroid[1], 0.0, 3)
+            msConfig = base.SingleFrameMeasurementConfig()
+            msConfig.algorithms.names = ["ext_shapeHSM_HsmPsfMoments"]
+            plugin, cat = makePluginAndCat(lsst.meas.extensions.shapeHSM.HsmPsfMomentsAlgorithm,
+                "ext_shapeHSM_HsmPsfMoments", centroid="centroid", control=lsst.meas.extensions.shapeHSM.HsmPsfMomentsControl())
```
Comment (TallJimbo) [inline tests/testHsm.py]: Too long.

File: ups/meas_extensions_shapeHSM.cfg [hunk 14]
```diff
@@ -3,7 +3,7 @@
 import lsst.sconsUtils
 
 dependencies = {
-    "required": ["utils", "afw", "meas_algorithms", "GalSim"],
+    "required": ["utils", "afw", "meas_base", "meas_algorithms", "GalSim"],
```
Comment (TallJimbo) [inline ups/meas_extensions_shapeHSM.cfg:6]: Can you actually remove meas_algorithms here (and below, in the .table file)?  Or does this package still use it somehow?

File: include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h (hunk 4 above)
Comment (jdswinbank) [inline include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h:114]: My assumption is that I'll wait until DM-980/1/2 all get merged, then update everything -- if that bumps my issue into S15, it won't worry me too much.

**Pull request merged**